bilby/gw/likelihood.py

```diff
@@ -1687,7 +1720,248 @@ def calculate_snrs_relative_binning(self, waveform_ratio_per_detector, interfero
+            # psd[:ifo_length][mask] = ifo.power_spectral_density_array[mask]
+            # d_inner_h += np.fft.fft(signal_long * data / psd)[in_prior]
+            # h_inner_h += ifo.optimal_snr_squared(signal=signal).real
+
+        if self.distance_marginalization:
+            time_log_like = self.distance_marginalized_likelihood(
+                d_inner_h, h_inner_h)
+        elif self.phase_marginalization:
+            time_log_like = (self._bessel_function_interped(abs(d_inner_h)) -
+                             h_inner_h.real / 2)
+        else:
+            time_log_like = (d_inner_h.real - h_inner_h.real / 2)
+
+        time_prior_array = self.priors['geocent_time'].prob(times)
+        time_post = (
+            np.exp(time_log_like - max(time_log_like)) * time_prior_array)
+
+        keep = (time_post > max(time_post) / 1000)
+        if sum(keep) < 3:
+            keep[1:-1] = keep[1:-1] | keep[2:] | keep[:-2]
+        time_post = time_post[keep]
+        times = times[keep]
+
+        new_time = Interped(times, time_post).sample()
+        return new_time
+
+    def generate_distance_sample_from_marginalized_likelihood(
+            self, waveform_ratio=None):
+        """
+        Generate a single sample from the posterior distribution for luminosity
+        distance when using a likelihood which explicitly marginalises over
+        distance.
+
+        See Eq. (C29-C32) of https://arxiv.org/abs/1809.02293
+
+        Parameters
+        ----------
+        signal_polarizations: dict, optional
+            Polarizations modes of the template.
+            Note: These are rescaled in place after the distance sample is
+                  generated to allow further parameter reconstruction to occur.
+
+        Returns
+        -------
+        new_distance: float
+            Sample from the distance posterior.
+        """
+        self.parameters.update(self.get_sky_frame_parameters())
+        if waveform_ratio is None:
+            waveform_ratio = \
+                self.compute_waveform_ratio(self.parameters)
+        d_inner_h, h_inner_h = self._calculate_inner_products(waveform_ratio)
+
+        d_inner_h_dist = (
+            d_inner_h * self.parameters['luminosity_distance'] /
+            self._distance_array)
+
+        h_inner_h_dist = (
+            h_inner_h * self.parameters['luminosity_distance']**2 /
+            self._distance_array**2)
+
+        if self.phase_marginalization:
+            distance_log_like = (
+                self._bessel_function_interped(abs(d_inner_h_dist)) -
+                h_inner_h_dist.real / 2)
+        else:
+            distance_log_like = (d_inner_h_dist.real - h_inner_h_dist.real / 2)
+
+        distance_post = (np.exp(distance_log_like - max(distance_log_like)) *
+                         self.distance_prior_array)
+
+        new_distance = Interped(
+            self._distance_array, distance_post).sample()
+
+        # self._rescale_signal(signal_polarizations, new_distance)
```

Comment:
This rescaling is commented out for now. But is it necessary to have this? If yes, any ideas on what we could do here? @adivijaykumar

Reply:
I guess we just need to rescale the waveform ratios. I might be wrong though. What do you think?

Reply:
But we don't use the rescaled waveforms anywhere, right?

Reply:
Yes, I think that is true. I am hence wondering what the motivation to rescale the signal is even in the exact case -- I strongly suspect it is the calculation of SNR. Could you check if that is the case?

Reply:
_Note: These are rescaled in place after the distance sample is generated to allow further parameter reconstruction to occur_
I found this in a docstring within `likelihood.py`

Reply:
![Screen Shot 2021-01-16 at 10 09 18 PM](https://user-images.githubusercontent.com/48632511/104817606-4bf26b00-5848-11eb-813e-1b895c0a141b.png)

I guess the current implementation is missing something especially in the phase and distance parts.

Reply:
> _Note: These are rescaled in place after the distance sample is generated to allow further parameter reconstruction to occur_
> I found this in a docstring within `likelihood.py`

This makes sense! For example if phase and distance are marginalized over along with time, this rescaling is necessary.


> I guess the current implementation is missing something especially in the phase and distance parts.
 Yeah, looks like it. Let me take a look later today.